code/network/lib/Loot/Network/ZMQ/Client.hs

```diff
@@ -191,37 +183,84 @@ data InternalRequest
     = IRUpdatePeers ZTUpdatePeersReq
     | IRRegister ClientId (Set MsgType) (Set Subscription) ZTClientEnv
     | IRReconnect (Set ZTNodeId)
+    deriving (Generic)
 
 instance T.Show InternalRequest where
     show (IRUpdatePeers r)      = "IRUpdatePeers: " <> show r
     show (IRRegister cId _ _ _) = "IRRegister " <> show cId
     show (IRReconnect nids)     = "IRReconnect" <> show nids
 
--- Get hashmap keys as a hashset.
-hmKeys :: HashMap k a -> HashSet k
-hmKeys = HSet.fromMap . HMap.map (const ())
+type CliRequestQueue = InternalQueue InternalRequest
+
+----------------------------------------------------------------------------
+-- Polling data
+----------------------------------------------------------------------------
 
--- | Wrapper over client request queue. CLient broker listens to this.
-newtype CliRequestQueue = CliRequestQueue
-    { unCliRequestQueue :: TQueue InternalRequest
+data CliPollData = CliPollData
+    { cpdPolls      :: ![Z.Poll Z.Socket IO]
+    , cpdN          :: !Int
+    , cpdResolveMap :: !(HashMap Int (ZTNodeId, PeerRes))
     }
 
+
+recreateCliPollData ::
+       PeersInfoVar -> CliRequestQueue -> ClientsQueue -> STM CliPollData
+recreateCliPollData piv reqQueue cQueue = do
+    peersInfo <- HMap.toList <$> readTVar piv
+    let cpdResolveMap :: HashMap Int (ZTNodeId, PeerRes)
+        cpdResolveMap = HMap.fromList $ [0..] `zip` peersInfo
+    let cpdN = HMap.size cpdResolveMap
+    let resources = map snd peersInfo
+    let toPoll sock = Z.Sock sock [Z.In] Nothing
+    let backPolls = map (toPoll . prBack) resources
+    let subPolls = map (toPoll . prSub) resources
+    let cliPolls = [toPoll (iqOut cQueue)]
+    let reqPolls = [toPoll (iqOut reqQueue)]
+    let cpdPolls = backPolls ++ subPolls ++ cliPolls ++ reqPolls
+    pure CliPollData{..}
+
+-- | This queue is used to transfer messages from clients to the main broker.
+type ClientsQueue = InternalQueue (Maybe ZTNodeId, (MsgType, Content))
+
+-- | This worker reads from clients' bitqueues and forwards these
+-- messages to one socket that is then processed by broker in one big
+-- zmq poll.
+clientsToBrokerWorker :: ZTNetCliEnv -> IO ()
+clientsToBrokerWorker ZTNetCliEnv { ztClientsQueue, ztClients, ztCliLogging } =
```

Comment:
nice copy-paste from `listenersToWorker`

does it make sense to abstract it in `Common.hs`

Reply:
Actually `lisetenersToWorker` is a copy-paste of `clientsToBroker`. Hmmm, it's hard to decide. I think generalising these two would only complicate the understanding. Workers are essentially three lines, but generalising them would require passing many parameters to the function. I guess I'd like to avoid that.